Fix crash of CES queue delay metric when CESTracker is nil

dlapcevic opened the pull request:
Avoid operator crashing from trying to access a field of a nil CESTracker. This can happen rarely with two consecutive CEP deletions that belong to the same CES. While the first CES update is being processed, the second CEP deletion causes the CES in cache to become empty, and CES update gets enqueued anyway, before the first CES update is finished. The first CES update then deletes CES tracker as the CES has 0 CEPs in it. Finally, the second CES update runs into the error.

Signed-off-by: Dorde Lapcevic <[dordel@google.com](mailto:dordel@google.com)>

alan-kut commented on operator/pkg/ciliumendpointslice/manager.go:
```diff
@@ -634,15 +634,17 @@ func (c *cesMgr) insertCESInWorkQueue(ces *cesTracker, baseDelay time.Duration)
-		// Reset the cesInsertedAt value
-		ces.cesInsertedAt = t
-		diff = timeSinceCESQueued.Seconds()
+	if ces, exists := c.desiredCESs.getCESTracker(cesName); exists {
```
```suggestion
	ces, exists := c.desiredCESs.getCESTracker(cesName)
	if !exists {
	  return
	}
```

alan-kut replied:
I'd suggest to change quickly return if ces doesn't exist to improve code structure, to make it more flat.

Here there is one ces but let say we would have 3 objects. The current code would look:
```
if obj1, exists := getObj1(); exist {
  some code
  if obj2, exists := getObj2(); exist {
    some code
    if obj3, exists := getObj3(); exist {
      some code
}
return
```
instead of
```
obj1, exists := getObj1()
if !exists {
  return
}
some code
obj2, exists := getObj2()
if !exists {
  return
}
some code
obj3, exists := getObj3()
if !exists {
  return
}
some code
```

dlapcevic replied:
I updated it as you suggested and made the same change in getCESMetricCountersAndClear() for consistency.

dlapcevic committed:
Fix crash of CES queue delay metric when CESTracker is nil

Avoid operator crashing from trying to access a field of a nil CESTracker. This can happen rarely with two consecutive CEP deletions that belong to the same CES. While the first CES update is being processed, the second CEP deletion causes the CES in cache to become empty, and CES update gets enqueued anyway, before the first CES update is finished. The first CES update then deletes CES tracker as the CES has 0 CEPs in it. Finally, the second CES update runs into the error.

Signed-off-by: Dorde Lapcevic <dordel@google.com>

sayboras reviewed (approved):
LGTM :100:

nebril reviewed (approved):
It would be nice to get some error logs for these conditions, as we will not know that this kind of error occurred, but I think it's OK to merge as is.
